pkg/importer/dataStream.go

```diff
@@ -324,121 +306,73 @@ func (d *DataStream) http() (io.ReadCloser, error) {
 	return countingReader, nil
 }
 
-//registryData - temporary data used by import from registry flow
-//is deleted as part of dataStream.Close()
-type registryData struct {
-	dataDir string
-	file    *os.File
-}
-
-func (r registryData) Cleanup() error {
-	glog.V(1).Infof("registryData - deleting all the data")
-	if r.file != nil {
-		r.file.Close()
-	}
-	err := os.RemoveAll(r.dataDir)
-	if err != nil {
-		glog.Errorf(" Failed removing directory ", err.Error())
-	}
-	return nil
-}
-
-const (
-	//ContainerDiskImageDir - Expected disk image location in container image as described in
-	//https://github.com/kubevirt/kubevirt/blob/master/docs/container-register-disks.md
-	ContainerDiskImageDir = "disk"
-	//TempContainerDiskDir - Temporary location for pulled containerImage
-	TempContainerDiskDir = "tmp"
-)
-
 //This import source downloads specified container image from registry location
 //Then it extracts the image to a temporary location and expects an image file to be located under /disk directory
 //If such exists it creates a Reader on it and returns it for further processing
 func (d *DataStream) registry() (io.ReadCloser, error) {
 
-	tmpData := registryData{"", nil}
-	tmpData.dataDir = filepath.Join(d.DataDir, TempContainerDiskDir)
-
-	imageDir := filepath.Join(tmpData.dataDir, ContainerDiskImageDir)
-
-	//1. create temporary directory if does not exist to which all the data will be extracted
-	if _, err := os.Stat(tmpData.dataDir); os.IsNotExist(err) {
-		err := os.Mkdir(tmpData.dataDir, os.ModeDir|os.ModePerm)
-		if err != nil {
-			glog.Errorf("Failed to create temporary directory")
-			return nil, errors.Wrapf(err, fmt.Sprintf("Failed to create tempdirectory %s", tmpData.dataDir))
-		}
+	if !d.HasScratchAvailable {
+		// No scratch space available, exit with code indicating we need scratch space.
+		d.RequiresScratch = true
+		return nil, errors.New("Scratch space required and none found")
 	}
 
-	//cleanup in case of failure
-	cleanup := true
-	defer func() {
-		if cleanup {
-			tmpData.Cleanup()
-		}
-	}()
+	imageDir := filepath.Join(d.DataDir, ContainerDiskImageDir)
```

Comment:
Yes, so the thought was the following flow:
1. use skopeo to get the container image in /scratch.
2. Extract the image layers to /data (clean up /scratch).
3. Find the disk image and return the reader to it.
4. In data stream, write the contents of the reader back to /scratch (clean up /data)
5. Use the qemu convert to write the end result to /data

This way there is never a point in which I have the image data on either pvc twice. However this assumes that /data is available which is not always the case (if block pvc). I wanted to avoid doing everything in scratch because I will have 2 copies of the data at some point (container image AND extracted layers). But I don't think I can avoid that.